### PR TITLE
fix: fix delay clean logic when multiple healthchecker was started

### PR DESCRIPTION
**Pull request by oowl**
When we have many healthchecker instance, if we need to delay clear anyone, it will add all `delay_purge` number to these target table, And then wait active healthcheck timer to check these number flag to execute real clear logic. Ideally, it's  work expected, but we have one for loop bug in clear logic. which will case delay clear logic just execute once for first healthchecker instance. So this will cause delay clear can not remove some target expectedly.

https://github.com/Kong/lua-resty-healthcheck/blob/master/lib/resty/healthcheck.lua#L1725

image

FTI-5478

### CODE REVIEW

**Comment by oowl** [comment]
It seems openresty has been remove old package from you deb package managment source, So i remove openresty 1.19 test
```
vagrant@ubuntu-focal:~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.6 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.6 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
vagrant@ubuntu-focal:~$ apt-cache madison openresty
 openresty | 1.21.4.3-1~focal1 | http://openresty.org/package/ubuntu focal/main amd64 Packages
 openresty | 1.21.4.2-1~focal1 | http://openresty.org/package/ubuntu focal/main amd64 Packages
 openresty | 1.21.4.1-1~focal1 | http://openresty.org/package/ubuntu focal/main amd64 Packages
```

File: lib/resty/healthcheck.lua [hunk 1]
```diff
@@ -1690,9 +1690,11 @@ function _M.new(opts)
         end
 
         local cur_time = ngx_now()
+        local is_checked = false
         for _, checker_obj in pairs(hcs) do
 
           if (last_cleanup_check + CLEANUP_INTERVAL) < cur_time then
+            is_checked = true
```
Comment (locao) [inline lib/resty/healthcheck.lua:1697]: I don't get how this change fixes the issue. It seems the same code for me. Did you get to consistently reproduce the bug?

File: t/with_worker-events/11-clear.t [hunk 2]
```diff
@@ -280,3 +280,86 @@ false
 target not found
 false
 target not found
+
+
+=== TEST 6: delayed_clear() would clear tgt list when we add two checkers
```
Comment (locao) [inline t/with_worker-events/11-clear.t:285]: Does this new test fail with the current `master` branch code?

File: .github/workflows/build_and_test_with_worker_events.yml [hunk 3]
```diff
@@ -19,7 +19,7 @@ jobs:
     runs-on: ubuntu-20.04
     strategy:
       matrix:
-        openresty-version: [1.19.9.1, 1.21.4.1]
+        openresty-version: [1.21.4.1]
```
Comment (locao) [inline .github/workflows/build_and_test_with_worker_events.yml:22]: Not a blocker for this PR, just a note for the feature: it's better to have different PRs for unrelated changes.

**Review by locao — CHANGES_REQUESTED** [review]
I have some questions before moving forward with this change.

File: t/with_worker-events/11-clear.t (hunk 2 above)
Comment (oowl) [reply t/with_worker-events/11-clear.t:285]: Yeah, I can make sure this test will be failed in our master branch.

File: lib/resty/healthcheck.lua (hunk 1 above)
Comment (oowl) [reply lib/resty/healthcheck.lua:1697]: Please be careful this LOC https://github.com/Kong/lua-resty-healthcheck/blob/master/lib/resty/healthcheck.lua#L1725

```
        for _, checker_obj in pairs(hcs) do

          if (last_cleanup_check + CLEANUP_INTERVAL) < cur_time then
            -- clear targets marked for delayed removal
            ****
            last_cleanup_check = cur_time
          end

          if checker_obj.checks.active.healthy.active and
            (checker_obj.checks.active.healthy.last_run +
              checker_obj.checks.active.healthy.interval <= cur_time)
          then
            checker_obj.checks.active.healthy.last_run = cur_time
            checker_callback(checker_obj, "healthy")
          end

          if checker_obj.checks.active.unhealthy.active and
            (checker_obj.checks.active.unhealthy.last_run +
              checker_obj.checks.active.unhealthy.interval <= cur_time)
          then
            checker_obj.checks.active.unhealthy.last_run = cur_time
            checker_callback(checker_obj, "unhealthy")
          end
        end
```

If we check the first healthcheck instance delayed removed target and replace `last_cleanup_check` with the current time. and then start checking the second healthcheck instance, it will skip because the first health check instance check logic has been executed `last_cleanup_check = cur_time`.

Comment (locao) [reply lib/resty/healthcheck.lua:1697]: Gotcha, it went over my head! Thanks for making it clear.

**Review by locao — APPROVED** [review]
🚀

**Review by outsinre — APPROVED** [review]

File: lib/resty/healthcheck.lua (hunk 1 above)
Comment (mcdullbloom) [reply lib/resty/healthcheck.lua:1697]: @oowl I got this problem too and I fixed it as you say.But there are targets left that not been cleared by delay clear when acquiring lock timeout.

```
2023/12/19 19:30:19 [error] 23583#23583: *467115 [lua] healthcheck.lua:1385: log(): [healthcheck] (upstream#upstreams/1) cleanup check failed, err: failed acquiring lock for 'lua-resty-healthcheck:upstream#upstreams/1:target_list_lock', timeout, context: ngx.timer, client: xxx, server: 0.0.0.0:80
```

the log to print the error

image

**Comment by ms2008** [comment]
@oowl Concerns from https://github.com/Kong/lua-resty-healthcheck/pull/146#discussion_r1432186363 will it happen?

**Comment by oowl** [comment]
@ms2008 check comment https://github.com/Kong/lua-resty-healthcheck/pull/146#discussion_r1429562364

**Comment by oowl** [comment]
> @oowl Concerns from [#146 (comment)](https://github.com/Kong/lua-resty-healthcheck/pull/146#discussion_r1432186363) will it happen?

Yeah, It will happen in some cases, you can run the newly added test in the master branch code.

File: lib/resty/healthcheck.lua (hunk 1 above)
Comment (oowl) [reply lib/resty/healthcheck.lua]: That's nice catch, I have resolved it. Please check again.

**Review by outsinre — APPROVED** [review]

**Review by windmgc — APPROVED** [review]
Code change LGTM

**Review by ms2008 — APPROVED** [review]
